File: BinaryTrees/max_level_sum.py
from collections import deque
from typing import Optional

from tree_node import TreeNode


class MaxLevelSum:
    def maxLevelSum(self, root: Optional[TreeNode]) -> int:
        ans = 0
        if not root:
            return ans

        queue = deque([root])
        level = 0
        max_sum = float("-inf")
        while queue:
            level += 1
            level_size = len(queue)
            level_sum = 0
            for _ in range(level_size):
                curr = queue.popleft()
                level_sum += curr.val
                if curr.left:
                    queue.append(curr.left)
                if curr.right:
                    queue.append(curr.right)
            if level_sum > max_sum:
                ans = level
                max_sum = level_sum
        return ans

    # Time complexity: O(n) where n is the number of nodes in the tree, 
    # since we visit each node exactly once.
    # Space complexity: O(m) where m is the maximum number of nodes at any level in the tree, 
    # which is the width of the tree.

File: BinaryTrees/maxLevelSum.js
const maxLevelSum = (root) => {
  let smallestMaxLevelSum = 0;
  let maxSum = -Infinity;
  let queue = [root];
  let currLevel = 0;
  while (queue.length) {
    currLevel++;
    const nextQueue = [];
    let levelSum = 0;
    for (const node of queue) {
      levelSum += node.val;
      if (node.left) nextQueue.push(node.left);
      if (node.right) nextQueue.push(node.right);
    }
    if (levelSum > maxSum) {
      maxSum = levelSum;
      smallestMaxLevelSum = currLevel;
    }
    queue = nextQueue;
  }
  return smallestMaxLevelSum;
};

// Time complexity: O(n) where n is the number of nodes in the tree, 
// since we visit each node exactly once.
// Space complexity: O(m) where m is the maximum number of nodes at any level in the tree (the width), 
// since we store the nodes of each level in a queue.